src/navigation/gate.py

```diff
@@ -24,111 +24,170 @@
+        Here is a reference for the points mentioned below:
+         https://github.com/umrover/mrover-ros/wiki/Navigation#searchtrajectory
+        :returns: an integer representing the farthest along point on the path that we can
+          drive to without intersecting the gate (while still driving through it)
         """
 
         # Get the shapes of both the posts
-        post_one_shape, post_two_shape = gate.get_post_shapes()
+        post_one_shape, post_two_shape = self.gate.get_post_shapes()
 
-        rover = rover_position[:2]
+        rover = self.rover_pos[:2]
 
         # try paths with successively more points until we have one that won't intersect
-        all_pts = np.vstack((prep, approach, center, done))
-        num_pts_included = 2
-        path = make_shapely_path(rover, all_pts[-num_pts_included:, :])
+        all_pts = self.__get_full_path()
+        num_pts_included = (
+            1 if self.passed_center else 2
+        )  # if we've already passed the center, we should not include it, otherwise we must to ensure we pass through the middle of the gate
+        path = self.__make_shapely_path(rover, all_pts[-num_pts_included:, :])
         while path.intersects(post_one_shape) or path.intersects(post_two_shape):
             num_pts_included += 1
             if num_pts_included == all_pts.shape[0]:
                 break
-            path = make_shapely_path(rover, all_pts[-num_pts_included:])
+            path = self.__make_shapely_path(rover, all_pts[-num_pts_included:])
 
-        coordinates = all_pts[-num_pts_included:]
-        coordinates = np.hstack((coordinates, np.zeros(coordinates.shape[0]).reshape(-1, 1)))
-        return coordinates
+        return all_pts.shape[0] - num_pts_included
 
+    def __make_shapely_path(self, rover, path_pts) -> LineString:
+        """
+        :param rover: position vector of the rover
+        :param pathPts: This is a np.array that has the coordinates of the path
+        :returns: Returns a Shapeley (geometry library) object of LineString which put the path points given into
+        one cohesive line segments.
+        """
+        path_list = np.vstack((rover, path_pts))
+        return LineString(path_list)
 
-def make_shapely_path(rover, path_pts) -> LineString:
-    """
-    :param rover: position vector of the rover
-    :param pathPts: This is a np.array that has the coordinates of the path
-    :returns: It returns the shapely object of LineString which put the path points given into
-    one cohesive line segments.
-    """
-    path_list = np.vstack((rover, path_pts))
-    return LineString(path_list)
+    def update(self, rover_pos: np.ndarray, gate: Gate) -> None:
+        """
+        Updates the calculated prep, approach, and center points based on the current gate estimate
+        """
+        self.rover_pos = rover_pos[:2]
+        self.gate = gate
+        self.__update_pts()
+
+    def get_cur_pt(self) -> Optional[np.ndarray]:
```

Comment:
Maybe make this function name more clear? Is "pt" used for "point" or "path"? The description makes it seem like this gets the current path but pt makes me think of a point.